torchrl/data/tensor_specs.py

```diff
@@ -1014,61 +1027,77 @@ class MultiDiscreteTensorSpec(DiscreteTensorSpec):
 
     def __init__(
         self,
-        nvec: Sequence[int],
+        nvec: Union[Sequence[int], torch.Tensor],
         device: Optional[DEVICE_TYPING] = None,
         dtype: Optional[Union[str, torch.dtype]] = torch.long,
     ):
+        if not isinstance(nvec, torch.Tensor):
+            nvec = torch.tensor(nvec)
+        self.nvec = nvec
         dtype, device = _default_dtype_and_device(dtype, device)
-        self._size = len(nvec)
-        shape = torch.Size([self._size])
-        space = BoxList([DiscreteBox(n) for n in nvec])
+        shape = nvec.size()
+        space = BoxList.from_nvec(nvec)
         super(DiscreteTensorSpec, self).__init__(
             shape, space, device, dtype, domain="discrete"
         )
 
+    def _rand(self, space: Box, shape: torch.Size):
+        x = []
+        for _s in space:
+            if isinstance(_s, BoxList):
+                x.append(self._rand(_s, shape))
+            else:
+                x.append(
+                    torch.randint(
+                        0,
+                        _s.n,
+                        torch.Size([1, *shape]),
+                        device=self.device,
+                        dtype=self.dtype,
+                    )
+                )
+        return torch.cat(x)
+
     def rand(self, shape: Optional[torch.Size] = None) -> torch.Tensor:
         if shape is None:
             shape = torch.Size([])
-        x = torch.cat(
-            [
-                torch.randint(
-                    0,
-                    space.n,
-                    torch.Size([1, *shape]),
-                    device=self.device,
-                    dtype=self.dtype,
-                )
-                for space in self.space
-            ]
-        ).squeeze()
-        _size = [self._size] if self._size > 1 else []
-        return x.T.reshape([*shape, *_size])
 
-    def _split(self, val: torch.Tensor):
-        return [val] if self._size < 2 else val.split(1, -1)
+        x = self._rand(self.space, shape)
+        _size = [] if self.shape == torch.Size([1]) else self.shape
+        return x.permute(*torch.arange(x.ndim - 1, -1, -1)).reshape([*shape, *_size])
 
     def _project(self, val: torch.Tensor) -> torch.Tensor:
         if val.dtype not in (torch.int, torch.long):
-            val = torch.round(val)
-        vals = self._split(val)
-        return torch.cat(
-            [
-                _val.clamp_(min=0, max=space.n - 1).unsqueeze(0)
-                for _val, space in zip(vals, self.space)
-            ],
-            dim=-1,
-        ).squeeze()
+            val = torch.round(val).type(self.dtype)
+        val = val.unsqueeze(0)
+        for permutation in itertools.product(*[range(d) for d in self.shape]):
```

Comment:
Can you explain what this does?
If it's to make sure that `val` with shape `[A, B, C, D]` is bounded by `self.nvec` with shape `[C, D]`, we can just do
```
maxvals = torch.randint(1, 10, (5, 6))
x = torch.randint(1, 10, (3, 4, 5, 6))
x[x >= maxvals] = maxvals.expand_as(x)[x >= maxvals] - 1
```

you can check that after that, all `x` are smaller than the corresponding `maxvals`

Reply:
You have understood what I have been trying to do. I didn't know that indexing in this way was possible. Thanks a lot for the tip, the code is much cleaner now!